filament/src/materials/antiAliasing/taa.mat

```diff
@@ -0,0 +1,280 @@
+    vec4 color = textureLodOffset(materialParams_color, uv.xy, 0.0, ivec2( 0,  0));
+#if FILTER_HISTORY
+    vec4 history = sampleTextureCatmullRom(materialParams_history, uv.zw,
+            vec2(textureSize(materialParams_history, 0)));
+#else
+    vec4 history = textureLod(materialParams_history, uv.zw, 0.0);
+#endif
+
+#if USE_YCoCg
+    history.rgb = RGB_YCoCg(history.rgb);
+#endif
+
+    // build the history clamping box
+    vec3 s[9];
+    s[0] = textureLodOffset(materialParams_color, uv.xy, 0.0, ivec2(-1, -1)).rgb;
+    s[1] = textureLodOffset(materialParams_color, uv.xy, 0.0, ivec2( 0, -1)).rgb;
+    s[2] = textureLodOffset(materialParams_color, uv.xy, 0.0, ivec2( 1, -1)).rgb;
+    s[3] = textureLodOffset(materialParams_color, uv.xy, 0.0, ivec2(-1,  0)).rgb;
+    s[4] = color.rgb;
+    s[5] = textureLodOffset(materialParams_color, uv.xy, 0.0, ivec2( 1,  0)).rgb;
+    s[6] = textureLodOffset(materialParams_color, uv.xy, 0.0, ivec2(-1,  1)).rgb;
+    s[7] = textureLodOffset(materialParams_color, uv.xy, 0.0, ivec2( 0,  1)).rgb;
+    s[8] = textureLodOffset(materialParams_color, uv.xy, 0.0, ivec2( 1,  1)).rgb;
+
+#if USE_YCoCg
+    s[0] = RGB_YCoCg(s[0]);
+    s[1] = RGB_YCoCg(s[1]);
+    s[2] = RGB_YCoCg(s[2]);
+    s[3] = RGB_YCoCg(s[3]);
+    s[4] = RGB_YCoCg(s[4]);
+    s[5] = RGB_YCoCg(s[5]);
+    s[6] = RGB_YCoCg(s[6]);
+    s[7] = RGB_YCoCg(s[7]);
+    s[8] = RGB_YCoCg(s[8]);
+    color.rgb = s[4].rgb;
+#endif
+
+    // unjitter/filter input
+    vec4 filtered = vec4(0, 0, 0, color.a);
+    for (int i = 0 ; i < 9 ; i++) {
+        filtered.rgb += s[i] * materialParams.filterWeights[i];
+    }
+
+    vec3 box5min = min(s[4], min(min(s[1], s[3]), min(s[5], s[7])));
+    vec3 box5max = max(s[4], max(max(s[1], s[3]), max(s[5], s[7])));
+    vec3 box9min = min(box5min, min(min(s[0], s[2]), min(s[6], s[8])));
+    vec3 box9max = max(box5max, max(max(s[0], s[2]), max(s[6], s[8])));
+    vec3 boxmin = (box5min + box9min) * 0.5;
+    vec3 boxmax = (box5max + box9max) * 0.5;
+
+    // TODO: filter history
+
+    // history clamping
+    history = clipToBox(BOX_CLIPPING, boxmin, boxmax, filtered, history);
+
+    float lumaColor   = luma(filtered.rgb);
+    float lumaHistory = luma(history.rgb);
+
+    float alpha = materialParams.alpha;
+#if PREVENT_FLICKERING
+    // [Lottes] prevents flickering by modulating the blend weight by the difference in luma
+    float diff = 1.0 - abs(lumaColor - lumaHistory) / (0.001 + max(lumaColor, lumaHistory));
+    alpha *= diff * diff;
+#endif
+
+    // tonemapping for handling HDR
+    filtered.rgb *= 1.0 / (1.0 + lumaColor);
+    history.rgb  *= 1.0 / (1.0 + lumaHistory);
+
+    // combine history and current frame
+    vec4 result = mix(history, filtered, alpha);
+
+    // untonemap result
+    result.rgb *= 1.0 / (1.0 - luma(result.rgb));
+
+#if USE_YCoCg
+    result.rgb = YCoCg_RGB(result.rgb);
+#endif
+
+    // store result (which will becomes new history)
+    result = max(vec4(0), result);
```

Comment:
Interesting, I'm surprised this could ever be negative going in.

Reply:
That's because of the bicubic/catmull-rom spline interpolation I think. Other implementations end up with the same fix.